File: program16.c
#include <stdio.h>

#define INF 9999
#define MAX 20

void dijkstra(int graph[MAX][MAX], int n, int start) {
    int cost[MAX][MAX], distance[MAX], visited[MAX], count, minDistance, nextNode, i, j;

    // Create cost matrix
    for (i = 0; i < n; i++)
        for (j = 0; j < n; j++)
            cost[i][j] = (graph[i][j] == 0) ? INF : graph[i][j];

    for (i = 0; i < n; i++) {
        distance[i] = cost[start][i];
        visited[i] = 0;
    }

    distance[start] = 0;
    visited[start] = 1;
    count = 1;

    while (count < n - 1) {
        minDistance = INF;

        for (i = 0; i < n; i++)
            if (distance[i] < minDistance && !visited[i]) {
                minDistance = distance[i];
                nextNode = i;
            }

        visited[nextNode] = 1;

        for (i = 0; i < n; i++)
            if (!visited[i])
                if (minDistance + cost[nextNode][i] < distance[i])
                    distance[i] = minDistance + cost[nextNode][i];

        count++;
    }

    printf("\nShortest distances from node %d:\n", start);
    for (i = 0; i < n; i++)
        if (i != start)
            printf("To %d = %d\n", i, distance[i]);
}

int main() {
    int graph[MAX][MAX], n, i, j, start;

    printf("Enter number of vertices: ");
    scanf("%d", &n);

    printf("Enter the adjacency matrix (0 for no edge):\n");
    for (i = 0; i < n; i++)
        for (j = 0; j < n; j++)
            scanf("%d", &graph[i][j]);

    printf("Enter the starting node: ");
    scanf("%d", &start);

    dijkstra(graph, n, start);

    return 0;
}

File: program16.py
import heapq

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = {}

    def addEdge(self, u, v, w):
        if u not in self.graph:
            self.graph[u] = []
        self.graph[u].append((v, w))

    def dijkstra(self, src):
        distances = {node: float('inf') for node in self.graph}
        distances[src] = 0

        pq = [(0, src)]  # priority queue: (distance, node)

        while pq:
            current_distance, current_node = heapq.heappop(pq)

            if current_distance > distances[current_node]:
                continue

            for neighbor, weight in self.graph[current_node]:
                distance = current_distance + weight

                if distance < distances[neighbor]:
                    distances[neighbor] = distance
                    heapq.heappush(pq, (distance, neighbor))

        print("\nShortest distances from source:", src)
        for node in distances:
            print(f"{src} -> {node} = {distances[node]}")


# ---------- MAIN PROGRAM ----------
vertices = int(input("Enter number of vertices: "))
g = Graph(vertices)

while True:
    print("\n--- Dijkstra's Algorithm Menu ---")
    print("1. Add weighted edge")
    print("2. Find shortest path")
    print("3. Exit")

    ch = int(input("Enter your choice: "))

    if ch == 1:
        u = input("Enter starting vertex: ")
        v = input("Enter ending vertex: ")
        w = int(input("Enter weight: "))
        g.addEdge(u, v, w)
        print("Edge added successfully!")

    elif ch == 2:
        src = input("Enter source vertex: ")
        g.dijkstra(src)

    elif ch == 3:
        print("Exiting...")
        break

    else:
        print("Invalid choice! Try again.")
